File: platforms/android/core/src/main/kotlin/io/adhush/core/RatingBug.kt
package io.adhush.core

import kotlin.math.max

/** Ported from config.RatingBugConfig; the ROI is relative to the found screen. */
data class RatingBugConfig(
    val roi: Roi = Roi(0.0, 0.02, 0.22, 0.16),
    val minFill: Double = 0.45,
    val minAspect: Double = 1.2,
    val maxAspect: Double = 4.0,
    val minShare: Double = 0.04,
    val maxShare: Double = 0.6,
    val appearAfterS: Double = 5.0,
    val minPresentS: Double = 1.0,
    val holdS: Double = 20.0,
)

/**
 * The parental-rating box as programme evidence (ADR 0023, after MythTV).
 * US networks flash "TV-14" in the upper left for ten seconds after every
 * break; commercials never carry one. Through the camera the box is a
 * bright, filled rectangle appearing in the corner of the found screen
 * where nothing bright was a moment before. Votes 0 with weight while it
 * is up and for [holdS] after; inert otherwise; never votes for an ad.
 * The screen box comes from the logo detector, which finds it every frame.
 */
class RatingBugDetector(
    private val screen: () -> Box?,
    private val cfg: RatingBugConfig = RatingBugConfig(),
) : Detector {
    override val name = "rating_bug"

    private var quietSince: Double? = null
    private var boxSince: Double? = null
    private var presentUntil = Double.NEGATIVE_INFINITY
    private var lastTs = 0.0
    var sightings = 0
        private set
    private var last = Triple(0.0, 0.0, 0.0)

    override fun warmup() { quietSince = null; boxSince = null; presentUntil = Double.NEGATIVE_INFINITY; sightings = 0 }

    override fun observeAudio(block: AudioBlock) {}

    /** (fill, aspect, share) of the bright bounding box inside the ROI. */
    fun boxScore(frame: Gray, roi: Box): Triple<Double, Double, Double> {
        var peak = 0f
        for (y in roi.y0 until roi.y1) for (x in roi.x0 until roi.x1) peak = max(peak, frame[x, y])
        val bright = max(140f, 0.75f * peak)
        var minX = Int.MAX_VALUE; var minY = Int.MAX_VALUE; var maxX = -1; var maxY = -1; var count = 0
        for (y in roi.y0 until roi.y1) for (x in roi.x0 until roi.x1) if (frame[x, y] >= bright) {
            count++; if (x < minX) minX = x; if (x > maxX) maxX = x; if (y < minY) minY = y; if (y > maxY) maxY = y
        }
        if (count < 8) return Triple(0.0, 0.0, 0.0)
        val w = (maxX - minX + 1).toDouble(); val h = (maxY - minY + 1).toDouble()
        return Triple(count / (w * h), w / h, (w * h) / (roi.w.toDouble() * roi.h))
    }

    override fun observeFrame(frame: Gray, ts: Double) {
        lastTs = ts
        val s = screen() ?: run { boxSince = null; return }
        val inScreen = cfg.roi.on(s.w, s.h)
        val roi = Box(s.x0 + inScreen.x0, s.y0 + inScreen.y0, s.x0 + inScreen.x1, s.y0 + inScreen.y1)
        if (roi.w < 4 || roi.h < 4 || roi.x1 > frame.w || roi.y1 > frame.h) return
        val (fill, aspect, share) = boxScore(frame, roi)
        last = Triple(fill, aspect, share)
        val boxed = fill >= cfg.minFill && aspect in cfg.minAspect..cfg.maxAspect && share in cfg.minShare..cfg.maxShare
        if (!boxed) { boxSince = null; if (quietSince == null) quietSince = ts; return }
        if (boxSince == null) {
            val q = quietSince
            if (q == null || ts - q < cfg.appearAfterS) return   // a permanent bright corner is scenery
            boxSince = ts; quietSince = null
        }
        if (ts - boxSince!! >= cfg.minPresentS) {
            if (presentUntil < ts) sightings++
            presentUntil = ts + cfg.holdS
        }
    }

    override val programPresent: Boolean get() = lastTs < presentUntil
    override val voting: Boolean get() = programPresent

    override fun vote(ts: Double): DetectorVote {
        val (fill, aspect, share) = last
        if (ts < presentUntil) return vote(ts, 0.0, "rating_box fill=${"%.2f".format(fill)} aspect=${"%.1f".format(aspect)} share=${"%.2f".format(share)} left_s=${"%.1f".format(presentUntil - ts)}")
        return vote(ts, 0.0, "no_rating_box fill=${"%.2f".format(fill)}")
    }

    fun describe(): String = "$sightings rating box(es) seen" + if (programPresent) " — programme" else ""
}

File: src/adhush/detect/rating_bug.py
"""The parental-rating box as an unmute cue (ADR 0023, after MythTV).

US networks flash the rating box — "TV-14", "TV-PG DL" — in the upper-left
corner for ten to fifteen seconds at the start of a programme and again
after every commercial break. Commercials never carry one. MythTV scores a
block for it; here it is *positive programme evidence*: the moment the box
appears, the programme is back, and the engine may end a mute early on it
(the late unmute being the failure this project fears most).

No OCR: the box is a bright, filled rectangle with dark text appearing
where nothing bright was a moment before. Inside the configured ROI the
detector finds the bounding box of bright pixels and asks whether it is
box-shaped (wider than tall, mostly filled, a good share of the ROI) and
whether it *appeared* — the ROI was not bright over the previous seconds —
and then stays for at least ``min_present_s``. It votes 0 (programme) with
weight while the box is up and for ``hold_s`` after, is inert otherwise, and
never votes for an ad.
"""

from __future__ import annotations

from typing import ClassVar

import numpy as np

from adhush.config import RatingBugConfig
from adhush.detect.base import Detector
from adhush.events import DetectorVote, FrameEvent
from adhush.util.imageops import downscale, extract_roi, to_luma

_TARGET_ROWS = 90


def box_score(roi: np.ndarray, bright: int) -> tuple[float, float, float]:
    """(fill, aspect, share) of the bright bounding box in a luma ROI:
    fill = bright pixels / box area, aspect = box w / h, share = box area /
    ROI area. All zero when nothing is bright."""
    mask = roi >= bright
    ys, xs = np.nonzero(mask)
    if len(ys) < 8:
        return 0.0, 0.0, 0.0
    h = int(ys.max() - ys.min() + 1)
    w = int(xs.max() - xs.min() + 1)
    area = float(h * w)
    return float(mask.sum()) / area, w / h, area / float(roi.size)


class RatingBugDetector(Detector):
    name: ClassVar[str] = "rating_bug"
    needs_video: ClassVar[bool] = True

    def __init__(self, config: RatingBugConfig) -> None:
        self._cfg = config
        self.warmup()

    def warmup(self) -> None:
        self._next_sample_ts = -np.inf
        self._quiet_since: float | None = None  # the ROI has had no box since
        self._box_since: float | None = None  # a box has been up since
        self._present_until = -np.inf
        self._last = (0.0, 0.0, 0.0)
        self._sightings = 0

    def _boxed(self, frame: np.ndarray) -> bool:
        luma = to_luma(frame)
        factor = max(1, luma.shape[0] // _TARGET_ROWS)
        small = downscale(luma, factor)
        r = self._cfg.roi
        roi = extract_roi(small, r.x, r.y, r.w, r.h)
        fill, aspect, share = box_score(roi, self._cfg.bright)
        self._last = (fill, aspect, share)
        return (
            fill >= self._cfg.min_fill
            and self._cfg.min_aspect <= aspect <= self._cfg.max_aspect
            and self._cfg.min_share <= share <= self._cfg.max_share
        )

    def observe_frame(self, event: FrameEvent) -> None:
        if event.ts < self._next_sample_ts:
            return
        self._next_sample_ts = event.ts + self._cfg.sample_interval_s
        boxed = self._boxed(event.frame)
        if not boxed:
            self._box_since = None
            if self._quiet_since is None:
                self._quiet_since = event.ts
            return
        # A box is up. It counts only if it *appeared*: the ROI had been
        # box-free for appear_after_s (a permanent bright corner is scenery).
        if self._box_since is None:
            if self._quiet_since is None or event.ts - self._quiet_since < self._cfg.appear_after_s:
                return
            self._box_since = event.ts
            self._quiet_since = None
        if event.ts - self._box_since >= self._cfg.min_present_s:
            if self._present_until < event.ts:
                self._sightings += 1
            self._present_until = event.ts + self._cfg.hold_s

    @property
    def program_present(self) -> bool:
        return self._next_sample_ts - self._cfg.sample_interval_s < self._present_until

    @property
    def voting(self) -> bool:
        return self.program_present

    def vote(self, ts: float) -> DetectorVote:
        fill, aspect, share = self._last
        if ts < self._present_until:
            return self._vote(ts, 0.0, f"rating_box fill={fill:.2f} aspect={aspect:.1f} share={share:.2f} left_s={self._present_until - ts:.1f}")
        return self._vote(ts, 0.0, f"no_rating_box fill={fill:.2f}")

    def describe(self) -> str:
        return f"{self._sightings} rating box(es) seen" + (" — programme" if self.program_present else "")
